aiokafka/conn.py

```diff
@@ -59,23 +105,40 @@ def _read(self):
 
                 resp = yield from self._reader.readexactly(size)
 
-                fut = self._requests.pop(0)
-                if not fut.cancelled():
-                    fut.set_result(resp)
+                recv_correlation_id, = self.HEADER.unpack(resp[:4])
+
+                correlation_id, resp_type, fut = self._requests.pop(0)
+                if (self._api_version == (0, 8, 2)
+                        and resp_type is GroupCoordinatorResponse
+                        and correlation_id != 0 and recv_correlation_id == 0):
+                    self.log.warning(
+                        'Kafka 0.8.2 quirk -- GroupCoordinatorResponse'
+                        ' coorelation id does not match request. This'
+                        ' should go away once at least one topic has been'
+                        ' initialized on the broker')
+
+                if correlation_id != recv_correlation_id:
+                    error = Errors.CorrelationIdError(
+                        'Correlation ids do not match: sent {}, recv {}'
+                        .format(correlation_id, recv_correlation_id))
+                    fut.set_exception(error)
+                    self.close()
+                    break
+
+                if not fut.done():
+                    response = resp_type.decode(resp[4:])
+                    self.log.debug('%s Response %d: %s',
+                                   self, correlation_id, response)
+                    fut.set_result(response)
         except OSError as exc:
-            conn_exc = ConnectionError("Kafka at {0}:{1} went away".format(
-                self._host, self._port))
+            conn_exc = Errors.ConnectionError(
+                "Connection at {0}:{1} broken".format(self._host, self._port))
             conn_exc.__cause__ = exc
             conn_exc.__context__ = exc
-            fut = self._requests.pop(0)
+            _, _, fut = self._requests.pop(0)
```

Comment:
Why do we only faise an error in the last request, not all of them on ConnectionError?